// @feathersjs/generators/lib/app/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generate = void 0;
const path_1 = require("path");
const chalk_1 = __importDefault(require("chalk"));
const pinion_1 = require("@feathershq/pinion");
const commons_1 = require("../commons");
const connection_1 = require("../connection");
const generate = (ctx) => (0, pinion_1.generator)(ctx)
    .then((0, commons_1.initializeBaseContext)())
    .then((ctx) => ({
    ...ctx,
    dependencies: [],
    devDependencies: []
}))
    .then((0, pinion_1.prompt)((ctx) => [
    {
        name: 'language',
        type: 'list',
        message: 'Do you want to use JavaScript or TypeScript?',
        when: !ctx.language,
        choices: [
            { name: 'TypeScript', value: 'ts' },
            { name: 'JavaScript', value: 'js' }
        ]
    },
    {
        name: 'name',
        type: 'input',
        when: !ctx.name,
        message: 'What is the name of your application?',
        default: ctx.cwd.split(path_1.sep).pop(),
        validate: (input) => {
            if (ctx.dependencyVersions[input]) {
                return `Application can not have the same name as a dependency`;
            }
            return true;
        }
    },
    {
        name: 'description',
        type: 'input',
        when: !ctx.description,
        message: 'Write a short description'
    },
    {
        type: 'list',
        name: 'framework',
        when: !ctx.framework,
        message: 'Which HTTP framework do you want to use?',
        choices: [
            { value: 'koa', name: `KoaJS ${chalk_1.default.grey('(recommended)')}` },
            { value: 'express', name: 'Express' }
        ]
    },
    {
        type: 'checkbox',
        name: 'transports',
        when: !ctx.transports,
        message: 'What APIs do you want to offer?',
        choices: [
            { value: 'rest', name: 'HTTP (REST)', checked: true },
            { value: 'websockets', name: 'Real-time', checked: true }
        ]
    },
    {
        name: 'packager',
        type: 'list',
        when: !ctx.packager,
        message: 'Which package manager are you using?',
        choices: [
            { value: 'npm', name: 'npm' },
            { value: 'yarn', name: 'Yarn' },
            { value: 'pnpm', name: 'pnpm' }
        ]
    },
    {
        name: 'client',
        type: 'confirm',
        when: ctx.client === undefined,
        message: (answers) => `Generate ${answers.language === 'ts' ? 'end-to-end typed ' : ''}client?`,
        suffix: chalk_1.default.grey(' Can be used with React, Angular, Vue, React Native, Node.js etc.')
    },
    {
        type: 'list',
        name: 'schema',
        when: !ctx.schema,
        message: 'What is your preferred schema (model) definition format?',
        suffix: chalk_1.default.grey(' Schemas allow to type, validate, secure and populate your data and configuration'),
        choices: [
            { value: 'typebox', name: `TypeBox ${chalk_1.default.grey('(recommended)')}` },
            { value: 'json', name: 'JSON schema' },
            { value: false, name: `No schema ${chalk_1.default.grey('(not recommended)')}` }
        ]
    },
    ...(0, connection_1.prompts)(ctx)
]))
    .then((0, pinion_1.runGenerators)(__dirname, 'templates'))
    .then((0, pinion_1.copyFiles)((0, pinion_1.fromFile)(__dirname, 'static'), (0, pinion_1.toFile)('.')))
    .then((0, commons_1.initializeBaseContext)())
    .then(async (ctx) => {
    const { dependencies } = await (0, connection_1.generate)(ctx);
    return {
        ...ctx,
        dependencies
    };
})
    .then((0, pinion_1.install)(({ transports, framework, dependencyVersions, dependencies, schema }) => {
    const hasSocketio = transports.includes('websockets');
    dependencies.push('@feathersjs/feathers', '@feathersjs/errors', '@feathersjs/schema', '@feathersjs/configuration', '@feathersjs/transport-commons', '@feathersjs/adapter-commons', '@feathersjs/authentication', '@feathersjs/authentication-client', 'winston');
    if (hasSocketio) {
        dependencies.push('@feathersjs/socketio');
    }
    if (framework === 'koa') {
        dependencies.push('@feathersjs/koa');
    }
    if (framework === 'express') {
        dependencies.push('@feathersjs/express', 'compression');
    }
    if (schema === 'typebox') {
        dependencies.push('@feathersjs/typebox');
    }
    return (0, commons_1.addVersions)(dependencies, dependencyVersions);
}, false, ({ packager }) => packager))
    .then((0, pinion_1.install)(({ language, devDependencies, dependencyVersions }) => {
    devDependencies.push('nodemon', 'axios', 'mocha', 'cross-env', 'prettier', '@feathersjs/cli', '@feathersjs/rest-client');
    if (language === 'ts') {
        devDependencies.push('@types/mocha', '@types/node', 'nodemon', 'ts-node', 'typescript', 'shx');
    }
    return (0, commons_1.addVersions)(devDependencies, dependencyVersions);
}, true, ({ packager }) => packager));
exports.generate = generate;
