// @feathersjs/generators/lib/service/templates/service.tpl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generate = exports.template = void 0;
const pinion_1 = require("@feathershq/pinion");
const commons_1 = require("../../commons");
const template = ({ camelName, authentication, isEntityService, path, lib, language, className, relative, schema, fileName }) => /* ts */ `// For more information about this file see https://dove.feathersjs.com/guides/cli/service.html
${authentication || isEntityService ? `import { authenticate } from '@feathersjs/authentication'` : ''}
${schema
    ? `
import { hooks as schemaHooks } from '@feathersjs/schema'
    
import {
  ${camelName}DataValidator,
  ${camelName}PatchValidator,
  ${camelName}QueryValidator,
  ${camelName}Resolver,
  ${camelName}ExternalResolver,
  ${camelName}DataResolver,
  ${camelName}PatchResolver,
  ${camelName}QueryResolver
} from './${fileName}.schema'
`
    : ''}

import type { Application } from '${relative}/declarations'
import { ${className}, getOptions } from './${fileName}.class'
${(0, commons_1.fileExists)(lib, `client.${language}`)
    ? `import { ${camelName}Path, ${camelName}Methods } from './${fileName}.shared'`
    : `
export const ${camelName}Path = '${path}'
export const ${camelName}Methods = ['find', 'get', 'create', 'patch', 'remove'] as const`}

export * from './${fileName}.class'
${schema ? `export * from './${fileName}.schema'` : ''}

// A configure function that registers the service and its hooks via \`app.configure\`
export const ${camelName} = (app: Application) => {
  // Register our service on the Feathers application
  app.use(${camelName}Path, new ${className}(getOptions(app)), {
    // A list of all methods this service exposes externally
    methods: ${camelName}Methods,
    // You can add additional custom events to be sent to clients here
    events: []
  })
  // Initialize hooks
  app.service(${camelName}Path).hooks({
    around: {
      all: [${authentication
    ? `
        authenticate('jwt'),`
    : ''} ${schema
    ? `
        schemaHooks.resolveExternal(${camelName}ExternalResolver),
        schemaHooks.resolveResult(${camelName}Resolver),`
    : ''}
      ],${isEntityService
    ? `
      find: [authenticate('jwt')],
      get: [authenticate('jwt')],
      create: [],
      update: [authenticate('jwt')],
      patch: [authenticate('jwt')],
      remove: [authenticate('jwt')]`
    : ''}
    },
    before: {
      all: [${schema
    ? `
        schemaHooks.validateQuery(${camelName}QueryValidator),
        schemaHooks.resolveQuery(${camelName}QueryResolver)
      `
    : ''}],
      find: [],
      get: [],
      create: [${schema
    ? `
        schemaHooks.validateData(${camelName}DataValidator),
        schemaHooks.resolveData(${camelName}DataResolver)
      `
    : ''}],
      patch: [${schema
    ? `
        schemaHooks.validateData(${camelName}PatchValidator),
        schemaHooks.resolveData(${camelName}PatchResolver)
      `
    : ''}],
      remove: []
    },
    after: {
      all: []
    },
    error: {
      all: []
    }
  })
}

// Add this service to the service type index
declare module '${relative}/declarations' {
  interface ServiceTypes {
    [${camelName}Path]: ${className}
  }
}
`;
exports.template = template;
const toServiceIndex = (0, pinion_1.toFile)(({ lib }) => [lib, 'services', `index`]);
const generate = (ctx) => (0, pinion_1.generator)(ctx)
    .then((0, commons_1.renderSource)(exports.template, (0, pinion_1.toFile)(({ lib, fileName, folder }) => [
    lib,
    'services',
    ...folder,
    `${fileName}`
])))
    .then((0, commons_1.injectSource)(({ camelName, folder, fileName }) => `import { ${camelName} } from './${folder.join('/')}/${fileName}'`, (0, pinion_1.prepend)(), toServiceIndex))
    .then((0, commons_1.injectSource)(({ camelName }) => `  app.configure(${camelName})`, (0, pinion_1.after)('export const services'), toServiceIndex));
exports.generate = generate;
